ui/raidboss/emulator/EmulatorCommon.js

```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+class EmulatorCommon {
+  static getTimestampFromLogLine(day, line) {
+    let result = EmulatorCommon.logLineRegex.exec(line);
+    return +new Date(day + ' ' + result.groups.lineTimestamp);
+  }
+}
+
+EmulatorCommon.logLineRegex = /^\[(?<lineTimestamp>[^\]]+)\] ([0-9A-Z]+):(.*)$/i;
+
+EmulatorCommon.SealRegexes = {
+  ja: /\[(?<lineTimestamp>[^\]]+)\] 00:0839:(?<Zone>.*)の封鎖まであと(?<Time>.*)秒/,
+  en: /\[(?<lineTimestamp>[^\]]+)\] 00:0839:(?<Zone>.*) will be sealed off in (?<Time>.*) seconds!/,
+  de: /\[(?<lineTimestamp>[^\]]+)\] 00:0839:Noch (?<Time>.*) Sekunden, bis sich (?<Zone>.*) schließt/,
+  // @TODO: Time? If we need it for some reason?
+  fr: /\[(?<lineTimestamp>[^\]]+)\] 00:0839:Fermeture (?<Zone>.*) dans /,
+  cn: /\[(?<lineTimestamp>[^\]]+)\] 00:0839:距(?<Zone>.*)被封锁还有(?<Time>.*)秒/,
+  ko: /\[(?<lineTimestamp>[^\]]+)\] 00:0839:(?<Time>.*)초 후에 (?<Zone>.*)(이|가) 봉쇄됩니다\./,
+};
+
+EmulatorCommon.EngageRegexes = {
+  ja: /\[(?<lineTimestamp>[^\]]+)\] 00:0039:戦闘開始！/,
+  en: /\[(?<lineTimestamp>[^\]]+)\] 00:0039:Engage!/,
+  de: /\[(?<lineTimestamp>[^\]]+)\] 00:0039:Start!/,
+  fr: /\[(?<lineTimestamp>[^\]]+)\] 00:0039:À l'attaque !/,
+  cn: /\[(?<lineTimestamp>[^\]]+)\] 00:0039:战斗开始！/,
+  ko: /\[(?<lineTimestamp>[^\]]+)\] 00:0039:전투 시작!/,
+};
+
+EmulatorCommon.CountdownRegexes = {
+  ja: /\[(?<lineTimestamp>[^\]]+)\] 00:.{4}:戦闘開始まで(?<Time>\d+)秒！/,
+  en: /\[(?<lineTimestamp>[^\]]+)\] 00:.{4}:Battle commencing in (?<Time>\d+) seconds!/,
+  de: /\[(?<lineTimestamp>[^\]]+)\] 00:.{4}:Noch (?<Time>\d+) Sekunden bis Kampfbeginn!/,
+  fr: /\[(?<lineTimestamp>[^\]]+)\] 00:.{4}:Début du combat dans (?<Time>\d+) secondes !/,
+  cn: /\[(?<lineTimestamp>[^\]]+)\] 00:.{4}:距离战斗开始还有(?<Time>\d+)秒！/,
+  ko: /\[(?<lineTimestamp>[^\]]+)\] 00:.{4}:전투 시작 (?<Time>\d+)초 전!/,
+};
+
+EmulatorCommon.EventDetailsRegexes = {
+  '03': /\[(?<lineTimestamp>[^\]]+)\] 03:(?<source_id>[^:]+):Added new combatant (?<source_name>[^:]+)\. {2}Job: (?<job>[^:]+) Level: (?<level>[^:]+) Max HP: (?<source_maxHP>\d+) Max MP: (?<source_maxMP>\d+) Pos: \((?<source_posX>[^,)]+),(?<source_posY>[^,)]+),(?<source_posZ>[^,)]+)\)/i,
```

Comment:
`resources/regexes.js` doesn't tend to capture all the fields. `03` in particular it only captures the ID and name.

Reply:
What do you think about adding them to `addedCombatantFull`?

Reply:
Also, if you need the timestamp on all of them, what about an additional `captureTimestamp` option? Or make `capture` return the timestamp as well?

Reply:
Probably the existing regexes should just have timestamp added to them. But right now I rely on some automatic mapping (pull log event type, then call `this['convert'+event]` if it exists, otherwise invoke a generic convert method) and I'd have to map every event ID to its `Regexes` entry.

There's also the issue of not knowing what language is in use ahead of time, which means for example having to have a copy of every `CactbotLanguage<code>` instantiated and hitting each of those separately for the seal/unseal/etc regexes, which is going to be pretty ugly.